File: app/src/main/java/info/horriblesubs/sher/fragment/ScheduleFragment.java
package info.horriblesubs.sher.fragment;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import info.horriblesubs.sher.R;
import info.horriblesubs.sher.adapter.ScheduleRecycler;
import info.horriblesubs.sher.model.base.ScheduleItem;
import info.horriblesubs.sher.model.response.ScheduleResponse;

public class ScheduleFragment extends Fragment {

    private static final String ARG_RESPONSE = "RESPONSE-SCHEDULE";
    private static final String ARG_NUMBER = "NUMBER-SCHEDULE";

    private ScheduleResponse scheduleResponse;
    private RecyclerView recyclerView;
    private int i;

    public static ScheduleFragment newInstance(ScheduleResponse scheduleResponse, int i) {
        ScheduleFragment fragment = new ScheduleFragment();
        Bundle args = new Bundle();
        args.putSerializable(ARG_RESPONSE, scheduleResponse);
        args.putInt(ARG_NUMBER, i);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_schedule, container, false);
        recyclerView = rootView.findViewById(R.id.recyclerView);
        return rootView;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        assert getArguments() != null;
        scheduleResponse = (ScheduleResponse) getArguments().getSerializable(ARG_RESPONSE);
        i = getArguments().getInt(ARG_NUMBER, 0);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        recyclerView.setItemAnimator(new DefaultItemAnimator());
        recyclerView.setLayoutManager(new GridLayoutManager(getContext(), 2));
        onLoadData();
    }

    private List<ScheduleItem> getSchedule() {
        List<ScheduleItem> scheduleItems = new ArrayList<>();
        Calendar calendar = Calendar.getInstance();
        for (ScheduleItem item : scheduleResponse.schedule) {
            calendar.setTime(item.getTime());
            if (calendar.get(Calendar.DAY_OF_WEEK) == i + 1 && item.isScheduled)
                scheduleItems.add(item);
            else if (i == 7 && !item.isScheduled)
                scheduleItems.add(item);
        }
        return scheduleItems;
    }

    private void onLoadData() {
        ScheduleRecycler scheduleRecycler = new ScheduleRecycler(getContext(), getSchedule());
        recyclerView.setAdapter(scheduleRecycler);
    }
}

File: app/src/main/java/info/horriblesubs/sher/ui/b/schedule/ScheduleFragment.kt
package info.horriblesubs.sher.ui.b.schedule

import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import info.horriblesubs.sher.R
import info.horriblesubs.sher.data.RepositoryResult
import info.horriblesubs.sher.data.subsplease.api.model.ItemSchedule
import info.horriblesubs.sher.databinding.BFragment3Binding
import info.horriblesubs.sher.functions.Info
import info.horriblesubs.sher.libs.dialog.InfoDialog
import info.horriblesubs.sher.libs.dialog.LoadingDialog
import info.horriblesubs.sher.libs.dialog.NetworkErrorDialog
import info.horriblesubs.sher.ui._extras.listeners.OnItemClickListener
import info.horriblesubs.sher.ui.b.MainActivity
import info.horriblesubs.sher.ui.c.ShowActivity
import info.horriblesubs.sher.ui.toast
import info.horriblesubs.sher.ui.viewBindings
import info.horriblesubs.sher.ui.viewModels
import java.util.*

class ScheduleFragment: Fragment(), OnItemClickListener<ItemSchedule.Show>,
    Toolbar.OnMenuItemClickListener {

    private val adapter by lazy { SchedulePagerAdapter(this, binding.viewPager, binding.tabLayout) }
    val model by viewModels<ScheduleModel>({requireActivity()})
    private var errorDialog: NetworkErrorDialog? = null
    private var loadingDialog: LoadingDialog? = null
    private lateinit var binding: BFragment3Binding
    private var infoDialog: InfoDialog? = null

    override fun onCreateView(inflater: LayoutInflater, group: ViewGroup?, bundle: Bundle?): View {
        binding = viewBindings(inflater, R.layout.b_fragment_3, group)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.toolbar.menu?.findItem(R.id.info)?.title = getString(R.string.about_weekly_schedule)
        model.resource.observe(viewLifecycleOwner) { onChanged(it) }
        errorDialog = context?.let{ NetworkErrorDialog(it) }
        binding.toolbar.setOnMenuItemClickListener(this)
        infoDialog = context?.let{ InfoDialog(it) }
    }

    private fun onChanged(t: RepositoryResult<ItemSchedule>?) = when(t?.status) {
        RepositoryResult.SUCCESS -> {
            adapter.result = t.value
            onSetLoading(false)
        }
        RepositoryResult.FAILURE -> onSetError(t.message)
        RepositoryResult.LOADING -> onSetLoading(true)
        else -> onSetError("Unexpected error occurred!!!")
    }

    private fun onSetError(str: String) {
        context.toast(str)
        onSetLoading(false)
        errorDialog?.show("https://subsplease.org/schedule")
    }

    private fun onSetLoading(b: Boolean) {
        loadingDialog = if (b) {
            onSetLoading(false)
            context?.let {
                LoadingDialog(it).apply {
                    show()
                }
            }
        } else {
            loadingDialog?.dismiss()
            null
        }
    }

    override fun onResume() {
        super.onResume()
        binding.viewPager.currentItem = Calendar.getInstance().get(Calendar.DAY_OF_WEEK) - 1
    }

    override fun onDestroyView() {
        model.resource.removeObservers(viewLifecycleOwner)
        loadingDialog?.dismiss()
        errorDialog?.dismiss()
        infoDialog?.dismiss()
        super.onDestroyView()
        model.stopServerCall
    }

    override fun onMenuItemClick(item: MenuItem?) = when(item?.itemId) {
        R.id.settings -> {
            (activity as? MainActivity)?.onNavigationItemSelected(item)
            true
        }
        R.id.refresh -> {
            model.refreshDataFromServer
            true
        }
        R.id.info -> {
            infoDialog?.show(Info.SCHEDULE, model.resourceTime.value) ?: Unit
            true
        }
        else -> false
    }

    override fun onItemClick(view: View, t: ItemSchedule.Show?, position: Int) {
        t?.page?.let { ShowActivity.startShowActivity(context, it) }
    }
}